src/main.cairo

```diff
@@ -1,32 +1,179 @@
 %lang starknet
 from starkware.cairo.common.math import assert_nn
 from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.starknet.common.syscalls import get_caller_address, get_block_timestamp, get_contract_address
+from starkware.cairo.common.math import assert_not_zero, assert_le
+
+from lib.cairo_contracts.src.openzeppelin.token.erc20.IERC20 import IERC20
+from src.interface.naming import Naming
+from src.interface.pricing import Pricing
+
+//
+// Storage
+//
+
+@storage_var
+func naming_contract() -> (contract_address: felt) {
+}
+
+@storage_var
+func pricing_contract() -> (contract_address: felt) {
+}
 
 @storage_var
-func balance() -> (res: felt) {
+func renews(user: felt, domain: felt) -> (bool: felt) {
+}
+
+@storage_var
+func voted_upgrade(user: felt, upgrade: felt) -> (bool: felt) {
+}
+
+//
+// Events
+//
+
+@event
+func ToggledRenewal(domain: felt, caller: felt, value: felt) {
+}
+
+@event
+func DomainRenewed(domain: felt, owner: felt, days: felt) {
+}
+
+@event
+func VotedUpgrade(caller: felt, upgrade: felt, vote: felt) {
+}
+
+@constructor
+func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    naming_address: felt, 
+    pricing_address: felt
+) {
+    naming_contract.write(naming_address);
+    pricing_contract.write(pricing_address);
+    return ();
 }
 
+// @notice Get the status of renewals for a domain
+// @param domain Domain to get status of
+// @param user User to get status of
+@view
+func is_renewing{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt, 
+    user: felt
+) -> (res: felt) {
+    let (res) = renews.read(user, domain);
+    return (res,);
+}
+
+// @notice Get the status of an upgrade vote
+// @param upgrade Upgrade to get status of
+// @param user User to get status of
+@view
+func has_voted_upgrade{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    user: felt,
+    upgrade: felt, 
+) -> (res: felt) {
+    let (res) = voted_upgrade.read(user, upgrade);
+    return (res,);
+}
+
+//
+// Externals
+//
+
+// @notice Activate or deactivate renewals for a domain
+// @param domain Domain to activate or deactivate renewals for
+// @dev callable by owner of domain
 @external
-func increase_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
-    amount: felt
+func toggle_renewals{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt,
 ) {
-    with_attr error_message("Amount must be positive. Got: {amount}.") {
-        assert_nn(amount);
+    alloc_locals;
+    let (caller) = get_caller_address();
+    let (contract) = naming_contract.read();
+    let (owner) = Naming.domain_to_address(contract, 1, cast(new(domain), felt*));
```

Comment:
The owner of a domain is not its target but the owner of its associated starknetID